packages/context/src/inject.ts

```diff
@@ -199,6 +216,31 @@ export function describeInjectedArguments(
   return meta || [];
 }
 
+function resolveByTag(ctx: Context, injection: Injection) {
+  const tag: string | RegExp = injection.metadata!.tag;
+  const bindings = ctx.findByTag(tag);
+  const values: BoundValue[] = new Array(bindings.length);
+
+  // A closure to set a value by index
+  const valSetter = (i: number) => (val: BoundValue) => (values[i] = val);
+
+  let asyncResolvers: PromiseLike<BoundValue>[] = [];
+  // tslint:disable-next-line:prefer-for-of
+  for (let i = 0; i < bindings.length; i++) {
+    const val = bindings[i].getValue(ctx);
+    if (isPromise(val)) {
+      asyncResolvers.push(val.then(valSetter(i)));
+    } else {
+      values[i] = val;
+    }
+  }
+  if (asyncResolvers.length) {
```

Comment:
Probably out of scope of this pull request: I feel we are repeating the `asyncResolvers` pattern in too many places. It would be good to extract a shared helper function.

A mock-up to illustrate my point:

```ts
function resolveByTag(ctx: Context, injection: Injection) {
  const tag: string | RegExp = injection.metadata!.tag;
  const bindings = ctx.findByTag(tag);
  
  return mapValuesOrPromises(bindings, b => b.getValue(ctx));
}

function mapValuesOrPromises<T, O>(
  list: ValueOrPromise<T>, 
  fn: (item: T) => ValueOrPromise<O>,
): ValueOrPromise<O[]> {
  const values: O[] = new Array(list.length);
  const asyncResolvers: PromiseLike<O>[] = [];

  const valSetter = (i: number) => (val: BoundValue) => (values[i] = val);
  
  // tslint:disable-next-line:prefer-for-of
  for (let i = 0; i < bindings.length; i++) {
    const val = fn(list[i]);
    if (isPromise(val)) {
      asyncResolvers.push(val.then(valSetter(i)));
    } else {
      values[i] = val;
    }
  }

  return asyncResolvers.length ?
    Promise.all(asyncResolvers).then(_ => values):
    values;
}
```

Reply:
Good point. See https://github.com/strongloop/loopback-next/pull/871